forest/src/paramfetch/mod.rs

```diff
@@ -0,0 +1,258 @@
+        Ok(()) => return Ok(()),
+        Err(e) => {
+            if e.kind() != ErrorKind::NotFound {
+                warn!("{}", e)
+            }
+        }
+    }
+
+    fetch_params(&path, info, mb).await?;
+
+    check_file(&path, info).map_err(|e| {
+        // TODO remove invalid file
+        e.into()
+    })
+}
+
+pin_project! {
+    struct FetchProgress<R> {
+        #[pin]
+        inner: R,
+        progress_bar: ProgressBar<pbr::Pipe>,
+    }
+}
+
+impl<R> FetchProgress<R> {
+    fn finish(&mut self) {
+        self.progress_bar.finish();
+    }
+}
+
+impl<R: AsyncRead + Unpin> AsyncRead for FetchProgress<R> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let r = Pin::new(&mut self.inner).poll_read(cx, buf);
+        if let Poll::Ready(Ok(size)) = r {
+            self.progress_bar.add(size as u64);
+        }
+        r
+    }
+}
+
+impl<R: BufRead + Unpin> BufRead for FetchProgress<R> {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<&'_ [u8]>> {
+        let this = self.project();
+        this.inner.poll_fill_buf(cx)
+    }
+
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        Pin::new(&mut self.inner).consume(amt)
+    }
+}
+
+async fn fetch_params(
+    path: &Path,
+    info: &ParameterData,
+    multi_bar: Option<Arc<MultiBar<Stdout>>>,
+) -> Result<(), Box<dyn StdError>> {
+    let gw = std::env::var(GATEWAY_ENV).unwrap_or_else(|_| GATEWAY.to_owned());
+    info!("Fetching {:?} from {}", path, gw);
+
+    let mut file = File::create(path).await?;
+
+    let url = format!("{}{}", gw, info.cid);
+
+    let client = Client::new();
+    let total_size: u64 = {
+        let res = client.head(&url).await?;
+        if res.status().is_success() {
+            res.header("Content-Length")
+                .and_then(|len| len.as_str().parse().ok())
+                .unwrap_or_default()
+        } else {
+            return Err(format!("failed to download file: {}", url).into());
+        }
+    };
+
+    let req = client.get(url.as_str());
+
+    if let Some(mb) = multi_bar {
+        let mut pb = mb.create_bar(total_size);
+        pb.set_units(Units::Bytes);
+
+        let mut source = FetchProgress {
+            inner: req.await?,
+            progress_bar: pb,
+        };
+        copy(&mut source, &mut file).await?;
+        source.finish();
+    } else {
+        let mut source = req.await?;
+        copy(&mut source, &mut file).await?;
+    };
+
+    Ok(())
+}
+
+fn check_file(path: &Path, info: &ParameterData) -> Result<(), io::Error> {
+    if std::env::var(TRUST_PARAMS_ENV) == Ok("1".to_owned()) {
+        warn!("Assuming parameter files are okay. DO NOT USE IN PRODUCTION");
+        return Ok(());
+    }
+
+    let mut file = SyncFile::open(path)?;
+    let mut hasher = Blake2b::new();
+    sync_copy(&mut file, &mut hasher)?;
```

Comment:
If someone has a good way to read from async buffer and write to sync buffer let me know lol (blake2b hasher doesn't have async write)